api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java

```diff
@@ -32,19 +31,41 @@
 
   BoundReference(Types.StructType struct, int fieldId) {
     this.fieldId = fieldId;
-    this.pos = find(fieldId, struct);
-    this.type = struct.fields().get(pos).type();
+    this.pos = findTopFieldPos(fieldId, struct);
+    this.type = (pos > -1) ? struct.fields().get(pos).type() : findStructFieldType(fieldId, struct);
   }
 
-  private int find(int fieldId, Types.StructType struct) {
+  private int findTopFieldPos(int fieldId, Types.StructType struct) {
     List<Types.NestedField> fields = struct.fields();
     for (int i = 0; i < fields.size(); i += 1) {
       if (fields.get(i).fieldId() == fieldId) {
         return i;
       }
     }
-    throw new ValidationException(
-        "Cannot find top-level field id %d in struct: %s", fieldId, struct);
+
+    return -1;
+  }
+
+  private Type findStructFieldType(int fieldId, Types.StructType struct) {
+    List<Types.NestedField> fields = struct.fields();
+    for (int i = 0; i < fields.size(); i += 1) {
+
+      if (fields.get(i).fieldId() == fieldId) {
+        return fields.get(i).type();
+      } else {
+
+        // look under struct by calling this method recursively
+        if (fields.get(i).type() instanceof Types.StructType) {
+          Types.StructType subStruct = fields.get(i).type().asStructType();
+
+          Type ret = findStructFieldType(fieldId, subStruct);
+          if (ret != null) {
+            return ret;
+          }
+        }
+      }
+    }
+    return null;
```

Comment:
Should this method be returning a null value in case none of the nested fields of this struct match by field id?
Instead of an instance of `BoundReference` with a null private attribute `type` would it be acceptable that we instead `throw new ValidationException("Cannot find nested field id %d in struct: %s", fieldId, struct);` ?

Reply:
done.